blockchain/spam_throttler.go

```diff
@@ -0,0 +1,303 @@
+var (
+	spamThrottler   *throttler = nil
+	spamThrottlerMu            = new(sync.RWMutex)
+)
+
+var (
+	thresholdGauge           = metrics.NewRegisteredGauge("txpool/throttler/threshold", nil)
+	candidateSizeGauge       = metrics.NewRegisteredGauge("txpool/throttler/candidate/size", nil)
+	throttledSizeGauge       = metrics.NewRegisteredGauge("txpool/throttler/throttled/size", nil)
+	allowedSizeGauge         = metrics.NewRegisteredGauge("txpool/throttler/allowed/size", nil)
+	throttlerUpdateTimeGauge = metrics.NewRegisteredGauge("txpool/throttler/update/time", nil)
+)
+
+type throttler struct {
+	config *ThrottlerConfig
+
+	candidates map[common.Address]int  // throttle candidates with spam weight. Not for concurrent use
+	throttled  map[common.Address]int  // throttled addresses. It requires mu.lock for concurrent use
+	allowed    map[common.Address]bool // white listed addresses. It requires mu.lock for concurrent use
+	mu         *sync.RWMutex           // mutex for throttled and allowed
+
+	threshold  int
+	throttleCh chan *types.Transaction
+	quitCh     chan struct{}
+}
+
+type ThrottlerConfig struct {
+	ActivateTxPoolSize uint `json:"activate_tx_pool_size"`
+	TargetFailRatio    uint `json:"target_fail_ratio"`
+	ThrottleTPS        uint `json:"throttle_tps"`
+	MaxCandidates      uint `json:"max_candidates"`
+
+	IncreaseWeight   int `json:"increase_weight"`
+	DecreaseWeight   int `json:"decrease_weight"`
+	InitialThreshold int `json:"initial_threshold"`
+	MinimumThreshold int `json:"minimum_threshold"`
+	ThrottleSeconds  int `json:"throttle_seconds"`
+}
+
+var DefaultSpamThrottlerConfig = &ThrottlerConfig{
+	ActivateTxPoolSize: 1000,
+	TargetFailRatio:    20,
+	ThrottleTPS:        100,   // len(throttleCh) = ThrottleTPS * 3 = 32KB * 10 * 5 = 960KB
+	MaxCandidates:      10000, // (20 + 4)B * 10000 = 240KB
+
+	IncreaseWeight:   5,
+	DecreaseWeight:   1,
+	InitialThreshold: 500,
+	MinimumThreshold: 100,
+	ThrottleSeconds:  300,
+}
+
+func GetSpamThrottler() *throttler {
+	spamThrottlerMu.RLock()
+	t := spamThrottler
+	spamThrottlerMu.RUnlock()
+	return t
+}
+
+func validateConfig(conf *ThrottlerConfig) error {
+	if conf == nil {
+		return errors.New("nil ThrottlerConfig")
+	}
+	if conf.TargetFailRatio > 100 {
+		return errors.New("invalid ThrottlerConfig. 0 <= TargetFailRatio <= 100")
+	}
+	if conf.MinimumThreshold <= conf.IncreaseWeight {
+		return errors.New("invalid ThrottlerConfig. IncreaseWeight < MinimumThreshold")
+	}
+	if conf.InitialThreshold < conf.MinimumThreshold {
+		return errors.New("invalid ThrottlerConfig. MinimumThreshold <= InitialThreshold")
+	}
+
+	return nil
+}
+
+// adjustThreshold adjusts the spam weight threshold of throttler in an adaptive way.
+func (t *throttler) adjustThreshold(ratio uint) {
+	var newThreshold int
+	// Decrease threshold if a fail ratio is bigger than target value to put more addresses in throttled map
+	if ratio > t.config.TargetFailRatio {
+		newThreshold = t.threshold - t.config.IncreaseWeight
```

Comment:
How about this code?
```suggestion
	if t.threshold > t.config.MinimumThreshold {	
		newThreshold = t.threshold - t.config.IncreaseWeight
	} else {	
	        newThreshold = t.config.MinimumThreshold
	}        
```

Reply:
Reflected. Please check this agains